src/lib.rs

```diff
@@ -251,16 +251,14 @@ entries but has not been committed yet, we must append the entries to the Raft l
     # }
     # let mut ready = node.ready();
     #
-    if !ready.entries().is_empty() {
-        // Append entries to the Raft log
-        node.mut_store().wl().append(ready.entries()).unwrap();
+    if let Some(hs) = ready.hs() {
+        // Raft HardState changed, and we need to persist it.
+        node.mut_store().wl().set_hardstate(hs.clone());
     }
-
     ```
 
-3. Check whether `hs` is empty or not. If not empty, it means that the `HardState` of the node has
-changed. For example, the node may vote for a new leader, or the commit index has been increased.
-We must persist the changed `HardState`:
+3. Check whether `entries` is empty or not. If not empty, it means that there are newly added
```

Comment:
Why handling hardstate first? If they are not written atomically, it's possible commit index in hard state pointing to an missing entry.

Reply:
I slipped up there. 
Is the order below ok?
```
1. messages
2. committed entries
3. snapshot
4. entries
5. hs
```

Reply:
Perhaps step 2 and 3 is conflict with each other. Or we should apply snapshot first and then committed entries to get a consistent apply order.

Reply:
Oh, I think it's ok because now `committed entries` and snapshot is exclusive now. 
https://github.com/tikv/raft-rs/blob/7aa651b9d83bc73221ff094da888753f6bdb5d92/src/raw_node.rs#L438-L454

Reply:
Actually they are not exclusive. For example, when leader sends a snapshot to a follower and then receives a heartbeat response from the same follower, leader will update the progress to probe and will send one entry to follower next time. So a follower is applying snapshot or handling snapshot doesn't mean it has no committed logs.